da/blob/verifier.go

```diff
@@ -46,35 +47,53 @@ type verifier[
 // newVerifier creates a new Verifier with the given proof verifier.
 func newVerifier[
 	BeaconBlockHeaderT BeaconBlockHeader[BeaconBlockHeaderT],
-	BlobSidecarT Sidecar[BeaconBlockHeaderT],
+	BlobSidecarT Sidecar[SignedBeaconBlockHeaderT],
 	BlobSidecarsT Sidecars[BlobSidecarT],
+	SignedBeaconBlockHeaderT SignedBeaconBlockHeader[BeaconBlockHeaderT],
 ](
 	proofVerifier kzg.BlobProofVerifier,
 	telemetrySink TelemetrySink,
-) *verifier[BeaconBlockHeaderT, BlobSidecarT, BlobSidecarsT] {
-	return &verifier[BeaconBlockHeaderT, BlobSidecarT, BlobSidecarsT]{
+) *verifier[
+	BeaconBlockHeaderT,
+	BlobSidecarT,
+	BlobSidecarsT,
+	SignedBeaconBlockHeaderT,
+] {
+	return &verifier[BeaconBlockHeaderT, BlobSidecarT,
+		BlobSidecarsT, SignedBeaconBlockHeaderT]{
 		proofVerifier: proofVerifier,
 		metrics:       newVerifierMetrics(telemetrySink),
 	}
 }
 
 // verifySidecars verifies the blobs for both inclusion as well
 // as the KZG proofs.
-func (bv *verifier[BeaconBlockHeaderT, _, BlobSidecarsT]) verifySidecars(
+func (bv *verifier[BeaconBlockHeaderT, _, BlobSidecarsT,
+	SignedBeaconBlockHeaderT],
+) verifySidecars(
 	sidecars BlobSidecarsT,
 	kzgOffset uint64,
 	blkHeader BeaconBlockHeaderT,
+	verifierFn func(
+		blkHeader BeaconBlockHeaderT,
+		signature crypto.BLSSignature,
+	) error,
 ) error {
 	defer bv.metrics.measureVerifySidecarsDuration(
 		time.Now(), math.U64(sidecars.Len()),
 		bv.proofVerifier.GetImplementation(),
 	)
 
-	// check that sideracs block headers match with header of the
+	// check that sidecars block headers match with header of the
 	// corresponding block
-	for i, s := range sidecars.GetSidecars() {
-		if !s.GetBeaconBlockHeader().Equals(blkHeader) {
-			return fmt.Errorf("unequal block header: idx: %d", i)
+	for _, s := range sidecars.GetSidecars() {
+		var sigHeader = s.GetSignedBeaconBlockHeader()
+		err := verifierFn(
```

Comment:
This removes any checks that validate that the `sigHeader` equals the `blkHeader`. This check could pass with any arbitrary values inside of `sigHeader` aside from the valid signature.

I believe we need to perform one of the two validations:
- ensure that `verifierFn(sigHeader, sigHeader.GetSignature()` also does not error
- ensure `sigHeader` == `blkHeader`

Reply:
Also note that the validation performed in this same function lower down can be removed after verification of equality of the two headers:
```go
	g.Go(func() error {
		return sidecars.ValidateBlockRoots()  // this can be removed
	})
```

Reply:
Totally agree.